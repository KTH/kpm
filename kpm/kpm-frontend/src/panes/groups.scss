@import "../vars";

#kpm-6cf53 .kpm-groups {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: calc(4 * var(--kpmGap));
  $colCount: 3;
  $colWidth: calc(100% / $colCount - $colGap * ($colCount - 1) / $colCount);

  & > ul {
    display: flex;
    flex-flow: row wrap;
    gap: $rowGap $colGap;
    align-items: flex-start;
    width: 100%;

    & > .kpm-groups-item {
      display: flex;
      align-items: start;
      width: $colWidth;
      margin-bottom: 0;
      padding: var(--kpmPad);
      gap: $rowGap $colGap;

      .icon.star {
        color: var(--kpmIconStarBg);
        // Fix alignment error
        margin-bottom: 2px;

        &.active {
          color: var(--kpmIconStarActiveBg);
        }
      }
    }
  }
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-groups > ul > .kpm-groups-item {
    flex-basis: 100%;
    font-size: 1.1em;
  }
}
