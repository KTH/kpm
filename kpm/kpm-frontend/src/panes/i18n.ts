export type TLang = "sv" | "en";

export function i18n(
  strObj: Record<TLang, string> | string,
  lang: TLang = "sv"
) {
  if (typeof strObj === "object") {
    return strObj[lang];
  }

  // TODO: Create language dict for static string lookup
  return strObj;
}
