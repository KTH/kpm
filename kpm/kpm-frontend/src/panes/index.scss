@use "./teaching.scss";

.kpm-studies-course {
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  gap: 0.5em;

  h2,
  h3 {
    font-size: 1em;
    margin: 0;
    padding: 0;
  }
}
