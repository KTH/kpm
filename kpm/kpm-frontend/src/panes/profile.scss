@use "../vars" as *;
@use "@kth/style/scss/tokens/typography";

#kpm-6cf53 div.kpm-profile {
  h3 {
    @include typography.font-heading-s;
    margin-block-start: 1rem;
    margin-block-end: 0.25rem;
  }

  h4 {
    @include typography.font-heading-xs;
    margin-block-start: 0.25rem;
    margin-block-end: 0.125rem;
  }
}
#kpm-6cf53 .kpm-profile .kpm-profile-link {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: $rowGap;

  display: flex;
  flex-flow: row nowrap;
  gap: $rowGap $colGap;
  align-items: center;

  img {
    height: 3em;
    width: 3em;
  }
}

#kpm-6cf53 .kpm-button.change-lang::after {
  display: inline-block;
  content: " ";
  width: 1em;
  height: 1em;
  background: $data-language-black-svg;
  margin: 0.2em 0 0.2em 1ch;
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-profile {
    & > .kpm-col {
      font-size: 1.1em;
    }
  }
}
