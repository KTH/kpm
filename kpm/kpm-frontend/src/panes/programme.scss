@use "../vars" as *;

#kpm-6cf53 .kpm-programs {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: calc(4 * var(--kpmGap));

  display: flex;
  flex-flow: column nowrap;
  
  .kpm-program-item {
    display: flex;
    flex-flow: row nowrap;
    gap: $rowGap $colGap;
    width: 100%;

    align-items: flex-start;
    padding: calc(2 * var(--kpmPad)) 0;
    border-bottom: 1px solid var(--kpmPaneSepatorFg);

    h2 {
      font-size: 1em;
      margin: 0;
      padding: 0;
    }

    .kpm-program-item-code {
      font-weight: 600;
    }

    .kpm-program-item-name {
      margin-right: auto;
    }
  }

}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-programs {
    .kpm-program-item {
      flex-wrap: wrap;
      a {
        width: calc(50% - 2 * var(--kpmGap));
      }
    }
  }
}
