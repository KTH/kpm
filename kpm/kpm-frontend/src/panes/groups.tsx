import React, { useState } from "react";
import { MenuPane, MenuPaneHeader } from "../components/menu";
import { APIGroups } from "kpm-backend-interface";
import { createApiUri, formatTerm, useDataFecther } from "./utils";
import {
  EmptyPlaceholder,
  ErrorMessage,
  LoadingPlaceholder,
} from "../components/common";
import { i18n } from "../i18n/i18n";
import "./groups.scss";
import { IconSettings, IconStar } from "../components/icons";
import { FilterOption, TabFilter } from "../components/filter";

export async function loaderStudies({ request }: any = {}): Promise<APIGroups> {
  const res = await fetch(createApiUri("/api/groups"), {
    signal: request?.signal,
  });
  const json = await res.json();
  return json;
}

type TFilter = "favs" | "all";

export function Groups() {
  const { res, loading, error } = useDataFecther<APIGroups>(loaderStudies);
  const { groups, group_search_url } = res || {};

  const [filter, setFilter] = useState<TFilter>("favs");

  const filteredGroups = groups?.filter((group) => {
    switch (filter) {
      case "favs":
        return group.starred;
      case "all":
        return true;
    }
  });

  const isEmpty = !loading && !error && filteredGroups?.length === 0;

  return (
    <MenuPane className="kpm-groups">
      <MenuPaneHeader title={i18n("My Groups")}>
        <a
          title="Search for interesting groups to follow"
          href="https://www.kth.se/search/?entityFilter=social-group&filterLabel=Groups"
        >
          {i18n("Find groups")}
        </a>
        <a href="https://www.kth.se/social/group-create/">
          {i18n("Create new group")}
        </a>
        <IconSettings href="https://www.kth.se/social/home/settings/groups" />
      </MenuPaneHeader>
      <TabFilter>
        <FilterOption<TFilter>
          value="favs"
          filter={filter}
          onSelect={setFilter}
        >
          {i18n("Favourites")}
        </FilterOption>
        <FilterOption<TFilter> value="all" filter={filter} onSelect={setFilter}>
          {i18n("All Subscriptions")}
        </FilterOption>
      </TabFilter>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {isEmpty && (
        <EmptyPlaceholder>
          {filter === "favs"
            ? i18n("You have no groups marked as favourites.")
            : i18n("You don't belong to any groups.")}
        </EmptyPlaceholder>
      )}
      {!isEmpty && (
        <ul>
          {filteredGroups?.map((group) => (
            <li>
              <IconStar className={group.starred ? "star active" : "star"} />
              <a href={group.url}>{group.name}</a>
            </li>
          ))}
        </ul>
      )}
    </MenuPane>
  );
}
