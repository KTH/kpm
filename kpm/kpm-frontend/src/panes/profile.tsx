import * as React from "react";
import { useLoaderData } from "react-router-dom";
import { ErrorMessage, LoadingPlaceholder } from "../components/common";
import { MenuPane } from "../components/menu";
import { createApiUri, useDataFecther } from "./utils";
import { i18n } from "../i18n/i18n";

export async function loaderProfile({ request }: any = {}) {
  const res = await fetch("/kpm/api", {
    signal: request?.signal,
  });
  const json = await res.json();
  return json;
}

export function Profile() {
  const { res, loading, error } = useDataFecther<{ msg: string }>(
    loaderProfile
  );
  const { msg } = res || {};
  // const { msg } = useLoaderData() as { msg: string };
  return (
    <MenuPane>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {msg && (
        <h2>
          {i18n("Profile")} {msg}
        </h2>
      )}
      {msg && (
        <p>
          <a href={createApiUri("/auth/logout")}>{i18n("Logout")}</a>
        </p>
      )}
    </MenuPane>
  );
}
