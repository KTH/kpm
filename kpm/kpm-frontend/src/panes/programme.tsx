import React, { useEffect, useState } from "react";
import { MenuPane, MenuPaneHeader } from "../components/menu";
import { APIProgrammes } from "kpm-backend-interface";
import { fetchApi, useDataFecther } from "./utils";
import {
  EmptyPlaceholder,
  ErrorMessage,
  LoadingPlaceholder,
} from "../components/common";
import { i18n } from "../i18n/i18n";
import "./programme.scss";
import { IconSettings, StarableItem } from "../components/icons";
import { FilterOption, TabFilter } from "../components/filter";

export async function loaderProgrammes({
  request,
}: any = {}): Promise<APIProgrammes> {
  const res = await fetchApi("/api/programmes", {
    signal: request?.signal,
  });
  const json = await res.json();
  if (res.ok) {
    return json;
  } else {
    // TODO: Handle more kinds of errors or keep it simple?
    throw new Error(json.message);
  }
}

export function useMutateProgrammes(res: APIProgrammes | undefined): {
  programmes: APIProgrammes["programmes"] | undefined;
  setStar(slug: string, value: boolean): void;
  errorSetStar: Error | undefined;
} {
  const [programmes, setProgrammes] = useState<APIProgrammes["programmes"]>();
  const [errorSetStar, setErrorSetStar] = useState<Error>();

  // Update state on res change (loaded/updated)
  useEffect(() => {
    if (res) {
      const { programmes } = res;
      setProgrammes(programmes);
      setErrorSetStar(undefined);
    }
  }, [res]);

  // ******************************************************
  // Allow widget to mutate the star property
  const setStar = async (slug: string, value: boolean) => {
    // Store to reset if call to API fails
    const oldProgrammes = programmes && [...programmes];
    // Clear error due to new interaction
    setErrorSetStar(undefined);

    let didChange = false;
    const newProgrammes = programmes?.map((programme) => {
      if (programme.slug === slug) {
        const { starred, ...other } = programme;
        if (programme.starred !== value) {
          didChange = true;
        }
        return {
          starred: value,
          ...other,
        };
      }
      return programme;
    });
    // Store change locally for quick feedback in UI
    setProgrammes(newProgrammes);

    if (didChange) {
      const res = await fetchApi("/api/star", {
        method: "post",
        body: JSON.stringify({
          kind: "program",
          slug,
          starred: value,
        }),
      }).catch((err: any) => {
        // Expose error and reset groups
        setErrorSetStar(err);
        setProgrammes(oldProgrammes);
      });
    }
  };

  return {
    programmes,
    setStar,
    errorSetStar,
  };
}

type TFilter = "favs" | "all";

export function Programme() {
  const { res, loading, error } =
    useDataFecther<APIProgrammes>(loaderProgrammes);
  const { programmes, setStar, errorSetStar } = useMutateProgrammes(res);

  const [filter, setFilter] = useState<TFilter>();

  // Switch to all if there are no starred programmes
  useEffect(() => {
    if (filter === undefined && Array.isArray(programmes)) {
      const hasStarred = !!programmes?.find((p) => p.starred);
      setFilter(hasStarred ? "favs" : "all");
    }
  }, [programmes]);

  const filteredProgrammes = programmes?.filter((programme) => {
    switch (filter) {
      case "favs":
        return programme.starred;
      case "all":
        return true;
    }
  });

  const isEmpty = !loading && !error && filteredProgrammes?.length === 0;

  return (
    <MenuPane className="kpm-programmes">
      <MenuPaneHeader title={i18n("My Programmes")}>
        <a
          title="Help / feedback for the personal menu in connection with the transition to new Ladok"
          href="https://www.kth.se/social/group/feedback-fran-anvand/page/personliga-menyn/"
        >
          Help / feedback
        </a>
        <IconSettings href="https://www.kth.se/social/home/settings/programmes" />
      </MenuPaneHeader>
      <TabFilter>
        <FilterOption<TFilter>
          value="favs"
          filter={filter || "all"}
          onSelect={setFilter}
        >
          {i18n("Favourites")}
        </FilterOption>
        <FilterOption<TFilter>
          value="all"
          filter={filter || "all"}
          onSelect={setFilter}
        >
          {i18n("All Subscriptions")}
        </FilterOption>
      </TabFilter>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {errorSetStar && <ErrorMessage error={errorSetStar} compact />}
      {isEmpty && (
        <EmptyPlaceholder>
          {filter === "favs"
            ? i18n("You have no programmes marked as favourites.")
            : i18n("You don't belong to any programme.")}
        </EmptyPlaceholder>
      )}
      {!isEmpty && (
        <ul>
          {filteredProgrammes?.map((programme) => (
            <StarableItem
              key={`kpm-program-${programme.slug}`}
              starred={programme.starred}
              onToggle={() => setStar(programme.slug, !programme.starred)}
            >
              <a href={programme.url}>{i18n(programme.name)}</a>
            </StarableItem>
          ))}
        </ul>
      )}
    </MenuPane>
  );
}
