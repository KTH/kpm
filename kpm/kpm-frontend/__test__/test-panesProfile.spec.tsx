import * as React from "react";
import { describe, expect, test } from "@jest/globals";
import renderer from "react-test-renderer";
import { RouterProvider, createMemoryRouter } from "react-router-dom";
import { Profile } from "../src/panes/profile";

describe("<Profile />", () => {
  test("Can be rendered", () => {
    const routes = [{ path: "/", element: <Profile /> }];
    const router = createMemoryRouter(routes, {
      initialEntries: ["/"],
      initialIndex: 1,
    });
    const component = renderer.create(<RouterProvider router={router} />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
