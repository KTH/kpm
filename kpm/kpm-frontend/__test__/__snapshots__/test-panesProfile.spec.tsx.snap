// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> Can be rendered 1`] = `
<dialog
  className="kpm-modal"
>
  <button
    className="kpm-modal-back-button kpm-mobile"
    onClick={[Function]}
  >
    Back to Personal Menu
  </button>
  <div
    className="kpm-modal-content kpm-profile"
  >
    <header
      className="kpm-modal-content-header"
    >
      <h2>
        Dr.LocalDev - Settings
      </h2>
      <aside
        className="kpm-modal-content-header-actions"
      >
        <button
          className="kpm-button change-lang"
          lang="sv"
          onClick={[Function]}
        >
          PÃ¥ svenska
        </button>
        <a
          className="kpm-button"
          href="/kpm/auth/logout"
        >
          Logout
        </a>
      </aside>
    </header>
    <div
      className="kpm-col"
    >
      <h3
        className="kpm-col-header"
      >
        About me
      </h3>
      <ul>
        <li>
          <a
            className="kpm-profile-link"
            href="https://www.kth.se/profile/jhsware/"
          >
            <img
              alt="Profile Image"
              src="https://www.kth.se/files/thumbnail/jhsware"
            />
            My Profile
          </a>
        </li>
        <li>
          <a
            href="https://www.kth.se/emailforward"
          >
            Forward KTH email
          </a>
        </li>
        <li>
          <a
            href="https://www.student.ladok.se/student/app/studentwebb/mina-uppgifter"
          >
            Contact details
          </a>
           
          (in LADOK)
        </li>
      </ul>
    </div>
    <div
      className="kpm-col"
    >
      <h3
        className="kpm-col-header"
      >
        My Settings
      </h3>
      <ul>
        <li>
          <a
            href="https://www.kth.se/social/home/settings/"
          >
            Subscriptions, my schedule
          </a>
        </li>
        <li>
          <a
            href="https://www.kth.se/social/home/calendar/settings/"
          >
            Export My schedule (e.g. to your mobile phone)
          </a>
        </li>
        <li>
          <a
            href="https://login.sys.kth.se/peap.html"
          >
            Wireless network
          </a>
        </li>
        <li>
          <a
            href="https://login.kth.se/password/change"
          >
            Change password
          </a>
        </li>
      </ul>
    </div>
  </div>
</dialog>
`;
